src/app/CommandSender.h

```diff
@@ -436,8 +510,49 @@ class CommandSender final : public Messaging::ExchangeDelegate
 
     CHIP_ERROR SendCommandRequestInternal(const SessionHandle & session, Optional<System::Clock::Timeout> timeout);
 
+    void OnResponseCallback(const ResponseData & aResponseData)
+    {
+        // mpExtendableCallback and mpCallback are mutually exclusive.
+        if (mpExtendableCallback)
+        {
+            mpExtendableCallback->OnResponse(this, aResponseData);
+        }
+        else if (mpCallback)
+        {
+            mpCallback->OnResponse(this, aResponseData.path, aResponseData.statusIB, aResponseData.data);
+        }
+    }
+
+    void OnErrorCallback(CHIP_ERROR aError)
+    {
+        // mpExtendableCallback and mpCallback are mutually exclusive.
+        if (mpExtendableCallback)
+        {
+            ErrorData errorData = { aError };
+            mpExtendableCallback->OnError(this, errorData);
+        }
+        else if (mpCallback)
+        {
+            mpCallback->OnError(this, aError);
+        }
+    }
+
+    void OnDoneCallback()
+    {
+        // mpExtendableCallback and mpCallback are mutually exclusive.
+        if (mpExtendableCallback)
+        {
+            mpExtendableCallback->OnDone(this);
+        }
+        else if (mpCallback)
+        {
+            mpCallback->OnDone(this);
+        }
+    }
+
     Messaging::ExchangeHolder mExchangeCtx;
     Callback * mpCallback                      = nullptr;
+    ExtendableCallback * mpExtendableCallback  = nullptr;
```

Comment:
Consider putting this together with `mpCallback` in a union to save RAM. An outer bitfield:1 could be used to differentiate which field is active.